application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java

```diff
@@ -252,22 +261,39 @@ public void upgradeRuleNodes() {
                                 ruleNode.setConfiguration(upgradeRuleNodeConfigurationResult.getSecond());
                             }
                             ruleNode.setConfigurationVersion(toVersion);
-                            ruleChainService.saveRuleNode(TenantId.SYS_TENANT_ID, ruleNode);
-                            log.debug("Successfully upgrade rule node with id: {} type: {} fromVersion: {} toVersion: {}",
-                                    ruleNodeId, ruleNodeTypeForLogs, fromVersion, toVersion);
+                            futures.add(jpaExecutorService.submit(() -> {
+                                ruleChainService.saveRuleNode(TenantId.SYS_TENANT_ID, ruleNode);
+                                log.debug("Successfully upgrade rule node with id: {} type: {} fromVersion: {} toVersion: {}",
+                                        ruleNodeId, ruleNodeTypeForLogs, fromVersion, toVersion);
+                            }));
+                            if (futures.size() >= MAX_PENDING_SAVE_RULE_NODE_FUTURES) {
+                                log.info("{} upgraded rule nodes so far ...",
+                                        totalRuleNodesUpgraded.addAndGet(awaitFuturesToCompleteAndGetCount(futures)));
+                                futures.clear();
+                            }
```

Comment:
At the end of the loop, you need to execute the same block again to wait for the final bunch of futures and print the total amount of futures that are processed.

Reply:
``` totalRuleNodesUpgraded``` does not need to be atomic since you update it in the same thread.

Reply:
> At the end of the loop, you need to execute the same block again to wait for the final bunch of futures and print the total amount of futures that are processed.

See line 281.

Reply:
> ` totalRuleNodesUpgraded` does not need to be atomic since you update it in the same thread.

It was set to atomic because of lambda forEach call: Variable used in lambda expression should be final or effectively final.